storage/allocator.go

```diff
@@ -13,75 +13,291 @@
 // permissions and limitations under the License. See the AUTHORS file
 // for names of contributors.
 //
-// Author: Levon Lloyd (levon.lloyd@gmail.com)
 // Author: Spencer Kimball (spencer.kimball@gmail.com)
+// Author: Kathy Spradlin (kathyspradlin@gmail.com)
 
 package storage
 
 import (
+	"fmt"
 	"math/rand"
+	"sort"
+	"sync"
 
+	"github.com/cockroachdb/cockroach/gossip"
 	"github.com/cockroachdb/cockroach/proto"
 	"github.com/cockroachdb/cockroach/util"
 )
 
-// allocator makes allocation decisions based on a zone configuration,
-// existing range metadata and available stores. Configuration
-// settings and range metadata information is stored directly in the
-// engine-backed range they describe. Information on suitability and
-// availability of servers is gleaned from the gossip network.
+const (
+	// minAvailCapacityThreshold: if the capacity of a store descriptor
+	// falls below this percentage, it will never be used as an
+	// rebalance target.
+	minAvailCapacityThreshold = 0.05
+	// maxAvailCapacityThreshold: if the mean available capacity of a
+	// list of store descriptors falls above this percentage, then the
+	// count of ranges on stores will be used to make rebalancing
+	// decisions instead of capacity.
+	maxAvailCapacityThreshold = 0.98
+	// rebalancePctFromMean: if the capacity or range count of a store
+	// is within rebalancePctFromMean of the mean, it is considered a
+	// viable target for rebalancing.
+	rebalancePctFromMean = 0.025 // 2.5%
+)
+
+// stat provides a running sample size and mean.
+type stat struct {
+	n, mean float64
+}
+
+// Update adds the specified value to the stat, augmenting the sample
+// size & mean.
+func (s *stat) Update(x float64) {
+	s.n += 1
+	s.mean += (x - s.mean) / s.n
+}
+
+// storeList keeps a list of store descriptors and associated count
+// and capacity stats across the stores.
+type storeList struct {
+	stores          []*proto.StoreDescriptor
+	count, capacity stat
+}
+
+// Update updates count and capacity stats using the specified descriptor.
+func (sl *storeList) Update(s *proto.StoreDescriptor) {
```

Comment:
nit: can the name and comment reflect that this method holds on to the descriptor? also, since `storeList` holds on to the descriptor (indefinitely?), the fact that `count` and `capacity` are `stat`s is an implementation detail, and should not be referenced in the doc comment